keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components/README.md

```diff
@@ -0,0 +1,239 @@
+  and reducing feature surface to improve quality.
+* --logtostderr, --alsologtostderr, --one-output, --stderrthreshold - 
+  responsible enabling/disabling writing to stderr (vs file). 
+  Motivation: Not needed if writing to files is removed.
+* --log-file-max-size, --skip-log-headers - responsible configuration of file 
+  rotation. 
+  Motivation: Not needed if writing to files is removed.
+* --add-dir-header, --skip-headers - klog format specific flags . 
+  Motivation: don't apply to other log formats
+* --log-backtrace-at - an legacy glog feature. 
+  Motivation: No trace of anyone using this feature.
+
+Flag deprecation should comply with standard k8s policy and require 3 releases before removal.
+
+This leaves that two flags that should be implemented by all log formats
+
+* -v - control global log verbosity of Info logs
+* --vmodule - control log verbosity of Info logs on per file level
+
+Those flags were chosen as they have direct effect of which logs are written, 
+directly impacting log volume and component performance.
+
+### User Stories
+
+#### Writing logs to files
+
+We should use go-runner as a official fallback for users that want to retain 
+writing logs to files. go-runner runs as parent process to components binary 
+reading it's stdout/stderr and is able to route them to files. go-runner is 
+already released as part of official K8s images it should be as simple as changing:
+
+```
+/usr/local/bin/kube-apiserver --log-file=/var/log/kube-apiserver.log
+```
+
+to
+
+```
+/go-runner --log-file=/var/log/kube-apiserver.log /usr/local/bin/kube-apiserver
+```
+
+### Caveats
+
+Is it ok for K8s components to drop support for subset of klog flags? 
+
+Technically K8s already doesn't support klog flags. Klog flags are renamed to 
+comply with K8s flag naming convention (underscores are replaced with hyphens).
+Full klog support was never promised to users and removal of those flags should
+be treated as removal of any other flag.
+
+Is it ok for K8s components to drop support writing to files?
+Writing directly to files is an important feature still used by users, but this
+doesn't directly necessitates direct support in components. By providing a
+external solution like go-runner we can allow community to develop more advanced
+features while maintaining high quality implementation within components.
+Having more extendable solution developed externally should be more beneficial
+to community when compared to forcing closed list of features on everyone.
+
+### Risks and Mitigations
+
+#### Users don't want to use go-runner as replacement.
+
+TODO
+
+#### Log processing in parent process causes performance problems
+
+TODO
+
+## Design Details
+
+TODO
+
+### Test Plan
+
+Go-runner is already used for scalability tests. We should ensure that we cover
+all existing klog features.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Klog can be configured without registering flags
+- Kubernetes logging configuration drops global state
```

Comment:
If there is no global logging state anymore, how will the current log calls like `klog.InfoS` be handled? That currently falls back to the globally configured klog.

I fully agree that we should get rid of that global state. I'd like to see a logger instance be passed into all functions which do logging, but there's no consensus on how to do that (attach to context vs. explicit parameter) and this KEP doesn't address this.

FWIW, I prefer the approach via context. Adding another parameter implies API breaks (client-go...) and other components already started using the context (logr, operator SDK), so if Kubernetes does the same, different code bases will become interoperable.

Reply:
I agree about using context to replace klog global state, but for now we want to start with removing configuration (flags etc), . This point is about configuration, basically removing global klog flags and passing through config struct to register/validate and initialize logging (same as how other components are configured). At some point this configuration will be applied to global klog state, but at least we clean up configuration.